Remote server discovery

Yashvardhan-A opened the pull request:
This PR adds the remote discovery feature to the reference app.

- Network discovery is performed through mDNS / DNS-SD using Android's NSD library.
- Spinner displaying discovered devices added to the startup and MandelbrotDemo activities.
- Dynamic addition of server and its devices performed using `pocl_remote_discovery_add_server()` function, see [PR](https://github.com/pocl/pocl/pull/1702).

This PR is to merged after the [Android: remote device addition](https://github.com/pocl/pocl/pull/1702) is merged to PoCL main.

Yashvardhan-A committed:
remote discovery

pjaaskel commented:
@RABijl  can you review&test this?

RABijl commented on android/app/src/main/java/org/portablecl/pocl_rreferenceandroidjavaclient/DNSDiscoveryHandler.java:
```diff
@@ -0,0 +1,122 @@
+                (NsdManager) activity.getApplicationContext().getSystemService(Context.NSD_SERVICE);
+        // Assign the RemoteDiscoveryManager instance
+        remoteDiscoveryManager = ds;
+        WifiManager wifiManager = (WifiManager) activity.getSystemService(Context.WIFI_SERVICE);
```
The wifimanager does not seem to be used.

RABijl commented on android/app/src/main/cpp/native-lib.cpp:
```diff
@@ -1,6 +1,7 @@
 #include <jni.h>
 #include <string>
 #include <android/log.h>
+#include "pocl_remote.h"
```
There should be a define to enable the new remote functions

RABijl commented on android/app/src/main/java/org/portablecl/pocl_rreferenceandroidjavaclient/RemoteDiscoveryManager.java:
```diff
@@ -0,0 +1,261 @@
+public class RemoteDiscoveryManager {
+
+    // Information about a discovered server
+    public class ServerInfo {
```
it would be nice to have a constructor for this

RABijl commented on android/app/src/main/java/org/portablecl/pocl_rreferenceandroidjavaclient/RemoteDiscoveryManager.java:
```diff
@@ -0,0 +1,261 @@
+    public static final String DEFAULT_SPINNER_LABEL = "Select a remote device";
+    private final DiscoverySpinnerAdapter deviceSpinnerAdapter;
+    public ArrayList<DeviceSelectionEntry> deviceList;
+    private final Spinner deviceSpinner;
```
the devicespinner is assigned, but not used outside the constructor

RABijl commented on android/app/src/main/java/org/portablecl/pocl_rreferenceandroidjavaclient/RemoteDiscoveryManager.java:
```diff
@@ -0,0 +1,261 @@
+    private void updateServerMap(String key, String sID, String txt, int deviceNum) {
+        // logic to decide if a discovered server is new or old, if server already exists, update it
+        if (discoveredServers.containsKey(key)) {
+            if (discoveredServers.get(key).serverID.equals(sID)) {
```
it is possible for discoveredservers to be null, which will cause an error.

RABijl commented on android/app/src/main/java/org/portablecl/pocl_rreferenceandroidjavaclient/DiscoverySpinnerAdapter.java:
```diff
@@ -0,0 +1,64 @@
+ */
+public class DiscoverySpinnerAdapter extends ArrayAdapter<RemoteDiscoveryManager.DeviceSelectionEntry> {
+
+    private Context context; // Context in which the adapter is used
```
context is assigned but not used

RABijl reviewed (commented):
Mostly looks good to me. There are a couple of small things That I've commented on and I will do a full text once https://github.com/pocl/pocl/pull/1702 has been merged.

RABijl commented on android/app/src/main/java/org/portablecl/pocl_rreferenceandroidjavaclient/StartupActivity.java:
```diff
@@ -74,13 +76,37 @@ public void onClick(View view) {
+                            Log.d("DISC", "Spinner position selected: " + position + " : server " +
+                                    "selected " +
+                                    ": " + selectedServer);
+                            remoteText.setText(selectedServer);
```
I've noticed that the remote device number gets added twice in the device info activity leading to things like: 
`<ip>:<port>/0/1`

pjaaskel commented:
@RABijl please do these small cleanups and send a new PR.

RABijl committed:
misc: clean up and update pocl

things fixed:
* native code defines remote extension
* removes defined but unused vars
* remove /0 string from device demo
* checks for null
* updates pocl to main

RABijl committed:
move info retrieval to own thread

Getting remote device info nowadays takes longer
than it used to, Therefore move a separate thread
to keep ui smooth.
